File: ios/CodexMeter/Infrastructure/DiagnosticLog.swift
import CodexMeterCore
import Foundation

/// Opt-in, privacy-filtered, rotating JSONL diagnostics stored only in app-private storage.
enum DiagnosticLog {
    static let unlockKey = "codex-meter.diagnostics-unlocked-v1"
    private static let enabledKey = "codex-meter.diagnostics-enabled-v1"
    private static let sessionKey = "codex-meter.diagnostics-session-v1"
    private static let directoryName = "diagnostic-logs"
    private static let currentFileName = "events.jsonl"
    private static let maxArchives = 2
    private static let maxFileBytes = 1_024 * 1_024
    private static let lock = NSLock()
    nonisolated(unsafe) private static var sequence: Int64 = 0

    struct Stats: Sendable, Equatable {
        var bytes: Int64
        var files: Int

        var hasLogs: Bool { bytes > 0 }
    }

    static func isUnlocked(defaults: UserDefaults = .standard) -> Bool {
        defaults.bool(forKey: unlockKey)
    }

    static func unlock(defaults: UserDefaults = .standard) {
        defaults.set(true, forKey: unlockKey)
    }

    static var isEnabled: Bool {
        UserDefaults.standard.bool(forKey: enabledKey)
    }

    static func setEnabled(_ enabled: Bool) {
        let wasEnabled = isEnabled
        guard wasEnabled != enabled else { return }
        if enabled {
            UserDefaults.standard.set(UUID().uuidString, forKey: sessionKey)
            UserDefaults.standard.set(true, forKey: enabledKey)
            info("diagnostics", "tracing_enabled", details: [
                "app_version": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "",
                "ios": ProcessInfo.processInfo.operatingSystemVersionString
            ])
        } else {
            info("diagnostics", "tracing_disabled")
            UserDefaults.standard.set(false, forKey: enabledKey)
        }
    }

    static func info(_ category: String, _ event: String, details: [String: String] = [:]) {
        write(level: "info", category: category, event: event, error: nil, details: details)
    }

    static func warn(_ category: String, _ event: String, details: [String: String] = [:]) {
        write(level: "warn", category: category, event: event, error: nil, details: details)
    }

    static func error(
        _ category: String,
        _ event: String,
        error: Error? = nil,
        details: [String: String] = [:]
    ) {
        write(level: "error", category: category, event: event, error: error, details: details)
    }

    static func stats() -> Stats {
        lock.lock()
        defer { lock.unlock() }
        var bytes: Int64 = 0
        var files = 0
        for file in orderedFiles() where FileManager.default.fileExists(atPath: file.path) {
            if let size = try? FileManager.default.attributesOfItem(atPath: file.path)[.size] as? NSNumber {
                let length = size.int64Value
                if length > 0 {
                    bytes += length
                    files += 1
                }
            }
        }
        return Stats(bytes: bytes, files: files)
    }

    static func clear() {
        lock.lock()
        defer { lock.unlock() }
        for file in orderedFiles() where FileManager.default.fileExists(atPath: file.path) {
            try? FileManager.default.removeItem(at: file)
        }
    }

    static func exportURL() throws -> URL {
        info("diagnostics", "export_started")
        lock.lock()
        defer { lock.unlock() }
        let export = FileManager.default.temporaryDirectory
            .appendingPathComponent("codex-meter-diagnostics.jsonl")
        if FileManager.default.fileExists(atPath: export.path) {
            try FileManager.default.removeItem(at: export)
        }
        FileManager.default.createFile(atPath: export.path, contents: nil)
        let handle = try FileHandle(forWritingTo: export)
        defer { try? handle.close() }
        for file in orderedFiles() {
            guard FileManager.default.fileExists(atPath: file.path),
                  let data = try? Data(contentsOf: file),
                  !data.isEmpty else {
                continue
            }
            try handle.write(contentsOf: data)
        }
        return export
    }

    static func formatBytes(_ bytes: Int64) -> String {
        if bytes < 1_024 {
            return "\(bytes) B"
        }
        if bytes < 1_024 * 1_024 {
            return "\(max(1, bytes / 1_024)) KB"
        }
        return String(format: "%.1f MB", Double(bytes) / (1_024.0 * 1_024.0))
    }

    private static func write(
        level: String,
        category: String,
        event: String,
        error: Error?,
        details: [String: String]
    ) {
        guard isEnabled else { return }
        do {
            var record: [String: Any] = [
                "timestamp": ISO8601DateFormatter().string(from: Date()),
                "sequence": nextSequence(),
                "session_id": UserDefaults.standard.string(forKey: sessionKey) ?? "",
                "level": DiagnosticSanitizer.redact(level),
                "category": DiagnosticSanitizer.redact(category),
                "event": DiagnosticSanitizer.redact(event),
                "thread": DiagnosticSanitizer.redact(Thread.current.name ?? "main")
            ]
            if !details.isEmpty {
                record["details"] = details.reduce(into: [String: String]()) { result, item in
                    result[DiagnosticSanitizer.redact(item.key)] = DiagnosticSanitizer.redact(item.value)
                }
            }
            if let error {
                record["error"] = [
                    "type": String(describing: type(of: error)),
                    "message": DiagnosticSanitizer.redact(error.localizedDescription)
                ]
            }
            let data = try JSONSerialization.data(withJSONObject: record, options: [.sortedKeys])
            var line = data
            line.append(contentsOf: "\n".utf8)
            append(line)
        } catch {
            // Diagnostics must never break the operation being diagnosed.
        }
    }

    private static func nextSequence() -> Int64 {
        lock.lock()
        defer { lock.unlock() }
        sequence += 1
        return sequence
    }

    private static func append(_ line: Data) {
        lock.lock()
        defer { lock.unlock() }
        let directory = directoryURL()
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let current = directory.appendingPathComponent(currentFileName)
        let existing = (try? FileManager.default.attributesOfItem(atPath: current.path)[.size] as? NSNumber)?
            .int64Value ?? 0
        if existing + Int64(line.count) > maxFileBytes {
            rotate(directory)
        }
        if !FileManager.default.fileExists(atPath: current.path) {
            FileManager.default.createFile(atPath: current.path, contents: nil)
        }
        guard let handle = try? FileHandle(forWritingTo: current) else { return }
        defer { try? handle.close() }
        _ = try? handle.seekToEnd()
        try? handle.write(contentsOf: line)
    }

    private static func rotate(_ directory: URL) {
        let oldest = archive(directory, index: maxArchives)
        if FileManager.default.fileExists(atPath: oldest.path) {
            try? FileManager.default.removeItem(at: oldest)
        }
        for index in stride(from: maxArchives - 1, through: 1, by: -1) {
            let source = archive(directory, index: index)
            if FileManager.default.fileExists(atPath: source.path) {
                try? FileManager.default.moveItem(at: source, to: archive(directory, index: index + 1))
            }
        }
        let current = directory.appendingPathComponent(currentFileName)
        if FileManager.default.fileExists(atPath: current.path) {
            try? FileManager.default.moveItem(at: current, to: archive(directory, index: 1))
        }
    }

    private static func orderedFiles() -> [URL] {
        let directory = directoryURL()
        return [
            archive(directory, index: 2),
            archive(directory, index: 1),
            directory.appendingPathComponent(currentFileName)
        ]
    }

    private static func archive(_ directory: URL, index: Int) -> URL {
        directory.appendingPathComponent("events.\(index).jsonl")
    }

    private static func directoryURL() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        return base
            .appendingPathComponent("CodexMeter", isDirectory: true)
            .appendingPathComponent(directoryName, isDirectory: true)
    }
}

File: android/app/src/main/java/dev/scorpion7slayer/modelsmeter/DiagnosticLog.java
package dev.scorpion7slayer.modelsmeter;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkCapabilities;
import android.net.Uri;
import android.os.Build;
import android.os.SystemClock;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import org.json.JSONArray;
import org.json.JSONObject;

/** Opt-in, privacy-filtered, rotating JSONL diagnostics stored only in app-private storage. */
public final class DiagnosticLog {
    private static final String DIRECTORY = "diagnostic-logs";
    private static final String CURRENT_FILE = "events.jsonl";
    private static final String PREF_ENABLED = "enabled";
    private static final String PREF_SESSION = "session_id";
    private static final String PREFERENCES = "models_meter_diagnostic_log_v1";
    private static final int MAX_ARCHIVES = 2;
    private static final long MAX_FILE_BYTES = 1024L * 1024L;
    private static final Object FILE_LOCK = new Object();
    private static final AtomicBoolean INSTALLED = new AtomicBoolean();
    private static final AtomicLong SEQUENCE = new AtomicLong();
    private static volatile Context applicationContext;

    public static final class Stats {
        public final long bytes;
        public final int files;

        Stats(long bytes, int files) {
            this.bytes = bytes;
            this.files = files;
        }

        public boolean hasLogs() {
            return bytes > 0L;
        }
    }

    private DiagnosticLog() {
    }

    public static void install(Context context) {
        Context app = appContext(context);
        if (app != null) {
            applicationContext = app;
        }
        if (app == null || !INSTALLED.compareAndSet(false, true)) {
            return;
        }
        if (isEnabled(app)) {
            preferences(app).edit().putString(PREF_SESSION, UUID.randomUUID().toString()).apply();
        }
        Thread.UncaughtExceptionHandler previous = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
            error(app, "process", "uncaught_exception", throwable,
                    "crashed_thread", thread == null ? "" : thread.getName());
            if (previous != null) {
                previous.uncaughtException(thread, throwable);
            }
        });
    }

    public static boolean isEnabled(Context context) {
        Context app = appContext(context);
        return app != null && preferences(app).getBoolean(PREF_ENABLED, false);
    }

    public static void setEnabled(Context context, boolean enabled) {
        Context app = appContext(context);
        if (app == null) {
            return;
        }
        install(app);
        boolean wasEnabled = isEnabled(app);
        if (wasEnabled == enabled) {
            return;
        }
        if (enabled) {
            String session = UUID.randomUUID().toString();
            preferences(app).edit()
                    .putString(PREF_SESSION, session)
                    .putBoolean(PREF_ENABLED, true)
                    .apply();
            info(app, "diagnostics", "tracing_enabled",
                    "app_version", appVersion(app),
                    "android_sdk", Build.VERSION.SDK_INT,
                    "device", Build.MANUFACTURER + " " + Build.MODEL);
        } else {
            info(app, "diagnostics", "tracing_disabled");
            preferences(app).edit().putBoolean(PREF_ENABLED, false).apply();
        }
    }

    public static void info(Context context, String category, String event, Object... fields) {
        write(context, "info", category, event, null, fields);
    }

    public static void warn(Context context, String category, String event, Object... fields) {
        write(context, "warn", category, event, null, fields);
    }

    public static void error(Context context, String category, String event, Throwable error,
            Object... fields) {
        write(context, "error", category, event, error, fields);
    }

    public static void info(String category, String event, Object... fields) {
        info(applicationContext, category, event, fields);
    }

    public static void warn(String category, String event, Object... fields) {
        warn(applicationContext, category, event, fields);
    }

    public static void error(String category, String event, Throwable error, Object... fields) {
        error(applicationContext, category, event, error, fields);
    }

    public static Stats stats(Context context) {
        Context app = appContext(context);
        if (app == null) {
            return new Stats(0L, 0);
        }
        synchronized (FILE_LOCK) {
            long bytes = 0L;
            int files = 0;
            for (File file : orderedFiles(app)) {
                if (file.isFile() && file.length() > 0L) {
                    bytes += file.length();
                    files++;
                }
            }
            return new Stats(bytes, files);
        }
    }

    public static void clear(Context context) {
        Context app = appContext(context);
        if (app == null) {
            return;
        }
        synchronized (FILE_LOCK) {
            for (File file : orderedFiles(app)) {
                if (file.exists()) {
                    file.delete();
                }
            }
        }
    }

    public static void export(Context context, Uri destination) throws Exception {
        Context app = appContext(context);
        if (app == null || destination == null) {
            throw new IllegalArgumentException("A diagnostic log destination is required.");
        }
        info(app, "diagnostics", "export_started");
        synchronized (FILE_LOCK) {
            try (OutputStream raw = app.getContentResolver().openOutputStream(destination, "wt")) {
                if (raw == null) {
                    throw new IllegalStateException("Android could not open the export file.");
                }
                try (BufferedOutputStream output = new BufferedOutputStream(raw)) {
                    byte[] buffer = new byte[16 * 1024];
                    for (File file : orderedFiles(app)) {
                        if (!file.isFile() || file.length() == 0L) {
                            continue;
                        }
                        try (BufferedInputStream input =
                                new BufferedInputStream(new FileInputStream(file))) {
                            int read;
                            while ((read = input.read(buffer)) != -1) {
                                output.write(buffer, 0, read);
                            }
                        }
                    }
                }
            }
        }
        info(app, "diagnostics", "export_finished");
    }

    public static String formatBytes(long bytes) {
        if (bytes < 1024L) {
            return bytes + " B";
        }
        if (bytes < 1024L * 1024L) {
            return Math.max(1L, bytes / 1024L) + " KB";
        }
        return String.format(java.util.Locale.US, "%.1f MB", bytes / (1024.0 * 1024.0));
    }

    private static void write(Context context, String level, String category, String event,
            Throwable error, Object... fields) {
        Context app = appContext(context);
        if (app == null || !isEnabled(app)) {
            return;
        }
        try {
            JSONObject record = new JSONObject();
            record.put("timestamp", Instant.now().toString());
            record.put("elapsed_ms", SystemClock.elapsedRealtime());
            record.put("sequence", SEQUENCE.incrementAndGet());
            record.put("session_id", preferences(app).getString(PREF_SESSION, ""));
            record.put("level", safe(level));
            record.put("category", safe(category));
            record.put("event", safe(event));
            record.put("thread", safe(Thread.currentThread().getName()));
            JSONObject details = details(fields);
            if (details.length() > 0) {
                record.put("details", details);
            }
            if ("network".equals(category)) {
                record.put("connectivity", connectivity(app));
            }
            if (error != null) {
                record.put("error", errorDetails(error));
            }
            append(app, record.toString() + "\n");
        } catch (Exception ignored) {
            // Diagnostics must never break the operation being diagnosed.
        }
    }

    private static JSONObject connectivity(Context app) throws Exception {
        JSONObject result = new JSONObject();
        try {
            ConnectivityManager manager =
                    (ConnectivityManager) app.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (manager == null) {
                result.put("available", false);
                return result;
            }
            NetworkCapabilities capabilities =
                    manager.getNetworkCapabilities(manager.getActiveNetwork());
            result.put("available", capabilities != null);
            result.put("metered", manager.isActiveNetworkMetered());
            if (capabilities != null) {
                result.put("internet",
                        capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET));
                result.put("validated",
                        capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED));
                result.put("wifi",
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI));
                result.put("cellular",
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR));
                result.put("ethernet",
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET));
                result.put("vpn",
                        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN));
            }
        } catch (RuntimeException exception) {
            result.put("available", false);
            result.put("read_error", safe(exception.getClass().getSimpleName()));
        }
        return result;
    }

    private static JSONObject details(Object... fields) throws Exception {
        JSONObject result = new JSONObject();
        if (fields == null) {
            return result;
        }
        for (int index = 0; index + 1 < fields.length; index += 2) {
            String key = safe(String.valueOf(fields[index]));
            Object value = fields[index + 1];
            if (value == null) {
                result.put(key, JSONObject.NULL);
            } else if (value instanceof Number || value instanceof Boolean) {
                result.put(key, value);
            } else {
                result.put(key, safe(String.valueOf(value)));
            }
        }
        return result;
    }

    private static JSONObject errorDetails(Throwable throwable) throws Exception {
        JSONObject result = new JSONObject();
        result.put("type", throwable.getClass().getName());
        result.put("message", safe(throwable.getMessage()));
        JSONArray stack = new JSONArray();
        StackTraceElement[] elements = throwable.getStackTrace();
        for (int index = 0; index < elements.length && index < 20; index++) {
            stack.put(safe(elements[index].toString()));
        }
        result.put("stack", stack);
        Throwable cause = throwable.getCause();
        if (cause != null && cause != throwable) {
            result.put("cause_type", cause.getClass().getName());
            result.put("cause_message", safe(cause.getMessage()));
        }
        return result;
    }

    private static void append(Context app, String line) throws Exception {
        byte[] encoded = line.getBytes(StandardCharsets.UTF_8);
        synchronized (FILE_LOCK) {
            File directory = directory(app);
            if (!directory.exists() && !directory.mkdirs()) {
                return;
            }
            File current = new File(directory, CURRENT_FILE);
            if (current.length() + encoded.length > MAX_FILE_BYTES) {
                rotate(directory);
            }
            try (FileOutputStream output = new FileOutputStream(current, true);
                    BufferedWriter writer = new BufferedWriter(
                            new OutputStreamWriter(output, StandardCharsets.UTF_8))) {
                writer.write(line);
            }
        }
    }

    private static void rotate(File directory) {
        File oldest = archive(directory, MAX_ARCHIVES);
        if (oldest.exists()) {
            oldest.delete();
        }
        for (int index = MAX_ARCHIVES - 1; index >= 1; index--) {
            File source = archive(directory, index);
            if (source.exists()) {
                source.renameTo(archive(directory, index + 1));
            }
        }
        File current = new File(directory, CURRENT_FILE);
        if (current.exists()) {
            current.renameTo(archive(directory, 1));
        }
    }

    private static File[] orderedFiles(Context app) {
        File directory = directory(app);
        return new File[] {
                archive(directory, 2),
                archive(directory, 1),
                new File(directory, CURRENT_FILE)
        };
    }

    private static File archive(File directory, int index) {
        return new File(directory, "events." + index + ".jsonl");
    }

    private static File directory(Context app) {
        return new File(app.getFilesDir(), DIRECTORY);
    }

    private static SharedPreferences preferences(Context app) {
        return app.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE);
    }

    private static Context appContext(Context context) {
        if (context == null) {
            return applicationContext;
        }
        Context app = context.getApplicationContext();
        return app == null ? context : app;
    }

    private static String safe(String value) {
        return DiagnosticSanitizer.redact(value == null ? "" : value);
    }

    private static String appVersion(Context app) {
        try {
            String version = app.getPackageManager()
                    .getPackageInfo(app.getPackageName(), 0).versionName;
            return version == null ? "" : version;
        } catch (Exception ignored) {
            return "";
        }
    }
}
